sphinx/source/docs/user_guide/server.rst

```diff
@@ -347,6 +336,82 @@ In this case you might have code similar to:
 
+.. _userguide_server_template:
+
+Customising the Application's Jinja Template
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As described above in :ref:`userguide_server_applications_directory`, you can override
+the default Jinja template used by the Bokeh server to generate the HTML code served to
+the user's browser.
+
+This opens up the possibility of managing the layout of the application in the client's
+browser using CSS, as well as making use of other Javascript libraries alongside BokehJS.
+
+See the `Jinja Project Documentation`_ for more details on how Jinja templating works.
+
+Embedding Figures in the Template
+'''''''''''''''''''''''''''''''''
+
+In the main thread of the Bokeh application, i.e. ``main.py``, any Bokeh figures
+that are going to be referenced in the templated code need to have their ``name``
+attribute set, and be added to the current document root.
+
+.. code-block:: python
+
+    from bokeh.plotting import curdoc
+
+    # Assuming a Bokeh Figure called bokeh_figure exists
+    bokeh_figure.name = "bokeh_jinja_figure"
+    curdoc().add_root(bokeh_figure)
+
+Then, in the corresponding Jinja template code, the figure may be referenced via the
+``roots`` template parameter, using the figure's ``name``, i.e.
+
+.. code-block:: html
+
+    {% extends base %}
+
+    {% block contents %}
+    <div>
+        {{ embed(roots.bokeh_jinja_figure) }}
+    </div>
+    {% endblock %}
+
+Defining Custom Variables
+'''''''''''''''''''''''''
+
+Custom variables can be passed to the template via the ``curdoc().template_variables``
+dictionary in place:
+
+.. code-block:: python
+
+    # set a new single key/value
+    curdoc().template_variables["user_id"] = user_id
+
+    # or update multiple at once
+    curdoc().template_variables.update(first_name="Mary", last_name="Jones")
+
+Custom Jinja templates and Threading
+''''''''''''''''''''''''''''''''''''
+It is possible for a connection between a client and server Document instance to
+timeout. Due to the nature of Bokeh's server-client communication, it is then possible
+that the client will reconnect to a server thread that doesn't know how to properly
+identify the Document's figure(s) to the client.
+
+In this scenario, the client will receive the Bokeh figures defined in the template anew,
+and will just append the figures to the end of the HTML body, in the order they are
+defined. This will likely break any HTML/CSS/Javascript code that expects the figure to
+be contained within specific HTML tags.
+
+If a Bokeh application that makes use of custom Jinja templates needs to scale beyond a
+single thread, then it is recommended to use multiple, single-threaded server instances
+with a load balancer to maintain consistency between the client and server. This is
+typically done using mechanisms such as session cookies.
+
+See :ref:`userguide_server_deployment_nginx_load_balance` for more details on running
+multiple server instances behind a load balancer.
```

Comment:
I'm confused by this section, as there should not be any reconnection at all after a timeout. Also I don't really know what you mean by multi-threaded in this context. Are you referring to `--num-procs`??

Reply:
> I'm confused by this section, as there should not be any reconnection at all after a timeout.

I might have misunderstood what is causing the problem - what I *thought* I had observed was the underlying websocket connection timing out, and then it reconnecting and getting all of the plots with new Div IDs.

I was able to reproduce this issue using both `--num-procs` with a value > 1 as well as with multiple Bokeh server instances behind a load balancer without session persistence.

> Also I don't really know what you mean by multi-threaded in this context. Are you referring to --num-procs??

That is what I meant - I'll clarify that multi-threaded means a value for `--num-procs` > 1.

Reply:
OK, I've revised this subsection to not go into the causes of the mismatch, more to identify when it may occur.